lib/inventory/controller.go

```diff
@@ -113,20 +157,33 @@ func NewController(auth Auth, opts ...ControllerOption) *Controller {
 
 	ctx, cancel := context.WithCancel(context.Background())
 	return &Controller{
-		store:            NewStore(),
-		serverKeepAlive:  options.serverKeepAlive,
-		serverTTL:        apidefaults.ServerAnnounceTTL,
-		maxKeepAliveErrs: options.maxKeepAliveErrs,
-		auth:             auth,
-		testEvents:       options.testEvents,
-		closeContext:     ctx,
-		cancel:           cancel,
+		store:              NewStore(),
+		serverKeepAlive:    options.serverKeepAlive,
+		serverTTL:          apidefaults.ServerAnnounceTTL,
+		instanceHBInterval: options.instanceHBInterval,
+		maxKeepAliveErrs:   options.maxKeepAliveErrs,
+		auth:               auth,
+		authID:             options.authID,
+		testEvents:         options.testEvents,
+		closeContext:       ctx,
```

Comment:
I'm not a huge fan of storing contexts like this. I'd prefer to pass context as an argument to any long running operations and have a separate close mechanism (like a close chan).

Up to you whether or not you want to address this. I know we do this all over the place so I don't mind if you choose not to make this change right now.

Reply:
Also not a huge fan of it, but I don't know of a good alternative.  We want in-flight heartbeat operations to succeed even if the client closes the stream during the write, but we want to cancel them immediately if the controller is closed.  I'm not aware of another mechanism to achieve this.

FWIW the main reason storing contexts in structs is bad is that contexts often times hold onto a bunch of request-scoped data, so storing them is a really common cause of memory leaks.  That isn't an issue here, since we use `context.Background()`.

Reply:
Yes, the other reason it comes a problem is it creates ambiguity as to how to close or shutdown the service, since the caller who passed in the original context may also hold a reference to it. 

I often find myself asking if I should call a Close method or shut down the context I originally passed in (or both!) to shut down anything still running in the background.

In this example, context is not passed in so I agree that this is okay.